exp/api/v1alpha3/azuremachinepool_webhook.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	capzv1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+)
+
+// log is for logging in this package.
+var azuremachinepoollog = logf.Log.WithName("azuremachinepool-resource")
+
+func (r *AzureMachinePool) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+
+// +kubebuilder:webhook:path=/mutate-exp-cluster-x-k8s-io-x-k8s-io-v1alpha3-azuremachinepool,mutating=true,failurePolicy=fail,matchPolicy=Equivalent,groups=exp.cluster.x-k8s.io.x-k8s.io,resources=azuremachinepools,verbs=create;update,versions=v1alpha3,name=mazuremachinepool.kb.io
+
+var _ webhook.Defaulter = &AzureMachinePool{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *AzureMachinePool) Default() {
+	azuremachinepoollog.Info("default", "name", r.Name)
+	if r.Spec.Template.Image == nil {
+
```

Comment:
in progress, or forgot a line?

Reply:
I forgot to remove this block. I started to implement the defaulting logic for images so that we didn't have to do it the reconciler. Upon digging into this, I asked @CecileRobertMichon why this was not done in `AzureMachine`. She was kind enough to remind me of a conversation we had previously about requesting another resource within a webhook and how we both agreed that it didn't seem right. 

The image defaulting logic is based on the version of k8s specified on the `Machine`. That means in the defaulter for `AzureMachine`, we would need to fetch `Machine`. Seemed kind of dirty, so we decided to keep it in the reconciler.

What do you think? It would be nice to centralize validation and defaulting logic.

Reply:
cc @vincepri who I also had a conversation with about this previously

Reply:
When we chatted about having defaulting in a webhook, I mentioned that it's way more preferable to not have any API calls done externally from the webhook code. 

Today, webhook deployments don't have RBAC permissions to make any calls, or have a Client that they can use. Moreover, having object inter-dependency in a webhook might complicate the defaulting logic, or in the worst case create loops of dependencies.

For these reason, I'd suggest to keep webhook defaulting logic as minimal as possible.

Reply:
It would be nice to be able to mirror a field from CAPI resources into the infra resource to get around something like this.

Reply:
@devigned Can you provide an example of what you'd expect to happen?

Reply:
An infra provider could create some sort of selector or expression that would contain a field / value combination from a related resource. The field / value would be used to assist with webhook mutations, like defaulting or validations. This would used to provide the webhook with information that is not available within a individual resource.

E.g. The `AzureMachine` hits the defaulting webhook. In an annotation, or some other place on the resource, we can find the K8s version from `Machine`, which will be used to determine the VM Image name and version used to build the VM. The webhook defaulter does not need to access any other resource to determine how to default the VM Image.